File: baekjoon/6549.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

long long n;
long long height[100001];
long long tree[1000001];
long long ans;

int init(int start, int end, int index) {
    if (start == end) {
        tree[index] = start;
        return tree[index];
    }
    int mid = (start + end) / 2;
    long long l = init(start, mid, index*2);
    long long r = init(mid+1, end, index*2+1);
    //cout << l << " " << r << "\n";
    tree[index] = height[l] > height[r] ? r : l;
    return tree[index];
}

int query(int start, int end, int index, int left, int right) {
    
    if (right < start || left > end) {
        return -1;
    }
    if (start >= left && end <= right) {
        return tree[index];
    }

    int mid = (start + end) / 2;
    long long l = query(start, mid, index*2, left, right);
    long long r = query(mid+1, end, index*2+1, left, right);
    if (l == -1) {
        return r;
    }
    else if (r == -1) {
        return l;
    }
    else {
        return height[r] > height[l] ? l : r;
    }
}

void dac(long long start, long long end) {

    if (start > end) return;
    int minIndex = query(0, n-1, 1, start, end);
    ans = max(ans, height[minIndex] * (end - start + 1));

    dac(start, minIndex-1);
    dac(minIndex+1, end);
}

int main()
{

    ios_base::sync_with_stdio(0);
    cin.tie(0);

    while (true)
    {
        ans = 0;
        cin >> n;
        if (n==0) break;
        for(int i=0; i<n; i++) {
            cin >> height[i];
        }
        init(0, n-1, 1);
        dac(0, n-1);
        cout << ans << "\n";
    }
}

File: _ongoing/6549.py
import sys, math
input = sys.stdin.readline

while True:
    arr = [*list(map(int, input().split()))]
    if arr[0] == 0:
        break
    tree = [0] * (arr[0] * 4)

    ans = -math.inf

    def init(start, end, index):
        """
        각 노드는 최소 높이의 인덱스를 저장한다.
        이는 후에 구간별 넓이를 계산할 때 최소 높이가 저장된 인덱스를 활용한다.
        """
        if start == end:
            tree[index] = start
            return tree[index]

        mid = (start + end) // 2
        l_min_idx = init(start, mid, index*2)
        r_min_idx = init(mid+1, end, index*2+1)
        tree[index] = arr[l_min_idx] if arr[l_min_idx] < arr[r_min_idx] else arr[r_min_idx]
        return tree[index]

    def query(start, end, index, left, right):
        if end < left or start > right:
            return -1
        if start >= left and end <= right:
            return tree[index]
        mid = (start + end) // 2
        l_min_idx = query(start, mid, index, left, right)
        r_min_idx = query(mid+1, end, index, left, right)
        if l_min_idx == -1:
            return r_min_idx
        elif r_min_idx == -1:
            return l_min_idx
        else:
            temp = arr[l_min_idx] if arr[l_min_idx] < arr[r_min_idx] else arr[r_min_idx]
            return temp

    def max_area(start, end):
        n = arr[0]
        m = query(start, end, 1, 1, n)

        area = (end-start+1)*arr[m]

        if start <= m-1:
            temp = max_area(start, m-1)
            if area < temp:
                area = temp
        
        if m+1 <= end:
            temp = max_area(m+1, end)
            if area < temp:
                area = temp

        return area

    init(1, arr[0], 1)
    print(max_area(1, arr[0]))
    


